File: userland/apps/textr/statbar.c
#include "textr.h"
#include <stdarg.h>
#include <stdio.h>
#include <string.h>

char *getModeStr(void) {
  switch (E.mode) {
  case MODE_INSERT:
    return "-- INSERT -- ";
    break;
  case MODE_COMMAND:
    return "-- COMMAND -- ";
    break;
  case MODE_VISUAL:
    return "-- VISUAL -- ";
    break;
  case MODE_NORMAL:
    return "-- NORMAL -- ";
    break;
  default:
    return "-- NORMAL -- ";
    break;
  }
}

void editorDrawStatusBar(WriteBuf *wb) {
  writeBufAppend(wb, "\x1b[7m", 4);
  char status[80], rstatus[80], mdstatus[20];

  int modelen = snprintf(
      mdstatus,
      sizeof(mdstatus),
      "%-10s",
      getModeStr());

  int statuslen = snprintf(
      status,
      sizeof(status),
      "%.20s - %d lines %s",
      E.buffer.filename ? E.buffer.filename : "[No Name]",
      E.buffer.num_rows,
      E.buffer.dirty ? "*" : "");

  int rlen = snprintf(
      rstatus,
      sizeof(rstatus),
      "%s | %d/%d",
      E.syntax ? E.syntax->filetype : "no ft",
      E.cursor.y + 1,
      E.buffer.num_rows);

  if (modelen > E.viewport.width) modelen = E.viewport.width;
  if (modelen + statuslen > E.viewport.width) statuslen = E.viewport.width - modelen;
  if (modelen + statuslen + rlen > E.viewport.width) rlen = E.viewport.width - modelen - statuslen;
  if (statuslen < 0) statuslen = 0;
  if (rlen < 0) rlen = 0;

  writeBufAppend(wb, mdstatus, modelen);
  writeBufAppend(wb, status, statuslen);

  int written = modelen + statuslen;
  while (written < E.viewport.width - rlen) {
    writeBufAppend(wb, " ", 1);
    written++;
  }
  writeBufAppend(wb, rstatus, rlen);
  writeBufAppend(wb, "\x1b[m", 3);
}

void editorSetStatusMsg(const char *fmt, ...) {
  va_list ap;
  va_start(ap, fmt);
  vsnprintf(E.ui.msg, sizeof(E.ui.msg), fmt, ap);
  va_end(ap);
  E.ui.msg_time = time(NULL);
}

void editorDrawMsgBar(WriteBuf *wb) {
  writeBufAppend(wb, "\x1b[K", 3);
  int msglen = strlen(E.ui.msg);
  if (msglen > E.viewport.width) msglen = E.viewport.width;
  if (msglen && time(NULL) - E.ui.msg_time < 5)
    writeBufAppend(wb, E.ui.msg, msglen);
}

File: userland/apps/asemics-rs/src/statbar.rs
use std::time::SystemTime;

use crate::consts::{LINE_NUM_RESERVE, SCREEN_CLEAR_LINE};
use crate::editor::{Editor, EditorMode};

impl Editor {
    pub fn mode_str(&self) -> &'static str {
        match self.mode {
            EditorMode::Insert => "-- INSERT -- ",
            EditorMode::Command => "-- COMMAND -- ",
            EditorMode::Visual => "-- VISUAL -- ",
            EditorMode::Normal => "-- NORMAL -- ",
        }
    }

    pub fn draw_status_bar(&self, wb: &mut Vec<u8>) {
        wb.extend_from_slice(b"\x1b[7m");

        let mdstatus = format!("{:<10}", self.mode_str());

        let status = format!(
            "{:.20} - {} lines {}",
            self.buffer.filename.as_deref().unwrap_or("[No Name]"),
            self.buffer.num_rows,
            if self.buffer.dirty { "*" } else { "" }
        );

        let rstatus = format!(
            "{} | {}/{}",
            self.syntax.map_or("no ft", |s| s.filetype),
            self.cursor.y + 1,
            self.buffer.num_rows
        );

        let mut modelen = mdstatus.len() as i32;
        let mut statuslen = status.len() as i32;
        let mut rlen = rstatus.len() as i32;

        if modelen > self.viewport.width {
            modelen = self.viewport.width;
        }
        if modelen + statuslen > self.viewport.width {
            statuslen = self.viewport.width - modelen;
        }
        if modelen + statuslen + rlen > self.viewport.width {
            rlen = self.viewport.width - modelen - statuslen;
        }
        if statuslen < 0 {
            statuslen = 0;
        }
        if rlen < 0 {
            rlen = 0;
        }

        wb.extend_from_slice(&mdstatus.as_bytes()[..modelen as usize]);
        wb.extend_from_slice(&status.as_bytes()[..statuslen as usize]);

        let mut written = modelen + statuslen;
        while written < self.viewport.width + LINE_NUM_RESERVE - rlen {
            wb.push(b' ');
            written += 1;
        }
        wb.extend_from_slice(&rstatus.as_bytes()[..rlen as usize]);
        wb.extend_from_slice(b"\x1b[m");
    }

    pub fn set_status_msg(&mut self, msg: String) {
        let mut msg = msg.into_bytes();
        msg.truncate(79); // C formats into a char msg[80]
        self.ui.msg = msg;
        self.ui.msg_time = SystemTime::now();
    }

    pub fn draw_msg_bar(&self, wb: &mut Vec<u8>) {
        wb.extend_from_slice(SCREEN_CLEAR_LINE);
        let mut msglen = self.ui.msg.len() as i32;
        if msglen > self.viewport.width {
            msglen = self.viewport.width;
        }
        let recent = self.ui.msg_time.elapsed().is_ok_and(|d| d.as_secs() < 5);
        if msglen > 0 && recent {
            wb.extend_from_slice(&self.ui.msg[..msglen as usize]);
        }
    }
}
